crates/spark/src/token/service.rs

```diff
@@ -0,0 +1,173 @@
+            })
+            .await?
+            .outputs_with_previous_transaction_data;
+        if outputs.is_empty() {
+            // Clear stored token outputs if none are returned
+            self.state.set_tokens_outputs(&[]).await?;
+            return Ok(());
+        }
+
+        // Raw token id to token outputs map
+        let mut outputs_map: HashMap<Vec<u8>, Vec<TokenOutputWithPrevOut>> = HashMap::new();
+        for output_with_previous_transaction_data in outputs {
+            let Some(output) = &output_with_previous_transaction_data.output else {
+                warn!("An empty output was returned from query_token_outputs, skipping");
+                continue;
+            };
+
+            let token_id = output.token_identifier().to_vec();
+            let token_output: TokenOutputWithPrevOut =
+                (output_with_previous_transaction_data, self.network).try_into()?;
+
+            outputs_map.entry(token_id).or_default().push(token_output);
+        }
+
+        // Fetch metadata for owned tokens
+        let token_identifiers = outputs_map.keys().cloned().collect();
+        let metadata = self.query_tokens_metadata(token_identifiers).await?;
+
+        if metadata.len() != outputs_map.keys().len() {
+            return Err(TokenOutputServiceError::Generic(
+                "Metadata not found for all tokens".to_string(),
+            ));
+        }
+
+        let token_outputs = outputs_map
+            .into_iter()
+            .map(|(token_id, outputs)| {
+                let metadata = metadata
+                    .iter()
+                    .find(|m| m.token_identifier == token_id)
+                    .ok_or_else(|| {
+                        TokenOutputServiceError::Generic("Metadata not found".to_string())
+                    })?;
+                let metadata = (metadata.clone(), self.network).try_into()?;
+                Ok(TokenOutputs { metadata, outputs })
+            })
+            .collect::<Result<Vec<TokenOutputs>, TokenOutputServiceError>>()?;
+
+        self.state.set_tokens_outputs(&token_outputs).await?;
+
+        Ok(())
+    }
+
+    async fn get_token_metadata(
+        &self,
+        filter: GetTokenOutputsFilter<'_>,
+    ) -> Result<Option<TokenMetadata>, TokenOutputServiceError> {
+        self.state
+            .get_token_outputs(filter)
+            .await
+            .map(|to| to.map(|to| to.metadata))
+    }
+
+    async fn insert_token_outputs(
+        &self,
+        token_outputs: &TokenOutputs,
+    ) -> Result<(), TokenOutputServiceError> {
+        self.state.insert_token_outputs(token_outputs).await
+    }
+
+    async fn reserve_token_outputs(
+        &self,
+        token_identifier: &str,
+        amount: u128,
+        preferred_outputs: Option<Vec<TokenOutputWithPrevOut>>,
+    ) -> Result<TokenOutputsReservation, TokenOutputServiceError> {
+        self.state
+            .reserve_token_outputs(token_identifier, amount, preferred_outputs)
```

Comment:
In select_leaves function of the TreeService we do some retries and refresh the leaves on first failure to mitigate sending right on start (before sync is completed). Do we want to apply a similar pattern here?